org.eclipse.xtext.util/src/org/eclipse/xtext/util/UriExtensions.java

```diff
@@ -8,35 +8,103 @@
 
+import com.google.inject.Singleton;
+
 /**
  * @author Christian Dietrich - Initial contribution and API
  * @since 2.14
  */
+@Singleton
 public class UriExtensions {
 	/**
-	 * Returns an URI with empty authority, if authority is absent (null) and has file scheme.
+	 * Converts the given absolute, encoded stringUri to an EMF URI. If the URI
+	 * is a {@link URI#isFile() file URI}, it is guaranteed that its
+	 * {@link URI#authority() authority} is not null.
 	 */
 	public URI toUri(String stringUri) {
-		try {
-			java.net.URI netUri = new java.net.URI(stringUri);
-			return toEmfUri(netUri);
-		} catch (URISyntaxException e) {
-			throw new RuntimeException(e);
+		URI result = URI.createURI(stringUri);
+		// Short-circuit the good case where the URI doesn't contain any unusual escape sequences
+		if (stringUri.contains("%")) {
+			return cleanupEncodedCharacters(result);
+		}
+		return withEmptyAuthority(result);
+	}
+
+	private URI cleanupEncodedCharacters(URI encoded) {
+		// Decompose the URI
+		String authority = encoded.authority();
+		String[] segments = encoded.segments();
+		String device = encoded.device();
+		String opaque = encoded.opaquePart();
+		String query = encoded.query();
+		String fragment = encoded.fragment();
+		
+		// Path the authority in-place if we are dealing with a file URI
+		if (authority == null && encoded.isFile()) {
+			authority = "";
+		} else if (authority != null && !authority.isEmpty()) {
+			authority = URI.encodeAuthority(URI.decode(authority), false);	
+		}
+		
+		// For encoded URIs, the device information might have been encoded as part of the segments
+		// so do a little dance if the device is absent and the first segment looks like one
+		int segmentIdx = 0;
+		if (device == null) {
+			if (segments.length > 0) {
+				String maybeDevice = URI.encodeSegment(URI.decode(segments[0]), false);
+				if (URI.validDevice(maybeDevice)) {
+					device = maybeDevice;
+					if (segments.length == 2 && segments[1].isEmpty()) {
+						segments = Strings.EMPTY_ARRAY;
+					} else if (segments.length > 1) {
+						segments = Arrays.copyOfRange(segments, 1, segments.length);
+					} else {
+						segments = null;
+					}
+				} else {
+					segments[0] = maybeDevice;
+					segmentIdx++;
+				}
+			}
+		}
+		// Handle segments, opaque-part, query, and fragment below
+		if (segments != null) {
+			for(int i = segmentIdx; i < segments.length; i++) {
+				segments[i] = URI.encodeSegment(URI.decode(segments[i]), false);
+			}
 		}
+		if (opaque != null) {
+			opaque = URI.encodeOpaquePart(URI.decode(opaque), false);
+		}
+		if (query != null) {
+			query = URI.encodeQuery(URI.decode(query), false);
+		}
+		if (fragment != null) {
+			query = URI.encodeFragment(URI.decode(fragment), false);
+		}
+		// Depending on the shape of the URI recreate it from the decomposed parts which are now
+		// correctly encoded in a minimal (only the strictly necessary but no optional escape sequences) way
```

Comment:
Is this the reason we re-encode the whole URI but the schema? Shall this comment rather be put as javadoc on the method?

Reply:
I'll make it a JavaDoc.

Yes, the cyrillic tests and the escape sequences in the device `c:` are the reason.